### PR TITLE
mayastor: Add flag for polling Reactor incoming thread queue

### LINKED ISSUE #438: Incoming queue polled to often
While running performance benchmarks, it observed that for each loop, we poll the reactor(s) for incoming threads. The queue on which we insert new threads is atomic which is rather expensive and could be polled less frequently. 

Instead of doing `while let Ok(i) = self.incomming.pop()` every iteration it is better to set a flag "incoming" to true when we schedule a new thread. When this flag is set, we should while loop over the queue.

### PR DESCRIPTION
**Pull request by daniel-shuy**
Closes #438 

Add `has_incoming` flag to `Reactor` to indicate that there are incoming threads to poll. This avoids polling the incoming thread queue too frequently.

### CODE REVIEW

File: mayastor/src/core/reactor.rs [hunk 1]
```diff
@@ -455,16 +460,19 @@ impl Reactor {
             t.poll();
         });
 
-        while let Ok(i) = self.incoming.pop() {
-            self.threads.borrow_mut().push_back(i);
+        if self.has_incoming.get() {
```
Comment (tiagolobocastro) [inline mayastor/src/core/reactor.rs:463]: I think this could be simplified into something like if self.has_incoming.replace(false) == true { ... }

**Review by tiagolobocastro — COMMENTED** [review]
Thanks for working on this @daniel-shuy !
This is where my rust knowledge is still not up to scratch yet but I wonder if has_incoming needs to be "volatile" here (or whatever the rust construct is), otherwise it could be optimized into a register since the compiler does not "see it change" in the polling loop?

File: mayastor/src/core/reactor.rs (hunk 1 above)
Comment (daniel-shuy) [reply mayastor/src/core/reactor.rs]: Done, thanks!

**Comment by daniel-shuy** [comment]
> This is where my rust knowledge is still not up to scratch yet but I wonder if has_incoming needs to be "volatile" here (or whatever the rust construct is), otherwise it could be optimized into a register since the compiler does not "see it change" in the polling loop?

@tiagolobocastro You're right, `Cell` is not thread-safe, I switched to using `AtomicBool` instead

**Comment by tiagolobocastro** [comment]
> > This is where my rust knowledge is still not up to scratch yet but I wonder if has_incoming needs to be "volatile" here (or whatever the rust construct is), otherwise it could be optimized into a register since the compiler does not "see it change" in the polling loop?
> 
> @tiagolobocastro You're right, `Cell` is not thread-safe, I switched to using `AtomicBool` instead.
> 
> `volatile` is not sufficient as it only prevents compiler optimization, but doesn't provide shared memory between threads, which means that `poll_times()` could read `has_incoming` as `false` even though it has been set to `true` in `schedule()`. `volatile` in Rust is only useful for accessing memory that can be modified outside the application (eg. memory-mapped I/O registers), is it useless for multithreading.

In this case the code is actually run by the same thread... but it's called through the `spdk_thread_create` "in C" and so the rust compiler might not have "visibility" over this which is why I suggested the volatile.

I did not raise this issue but as I understand it, what we're trying to avoid here is the cost of polling the atomic incoming queue and thus why we're trying to avoid atomics here so we just need the compiler to check the actual memory everytime (changed by the same thread) and not "cache" the flag in a register.

**Comment by gila** [comment]
The original patch -- i.e. without the atomic bool was indeed correct. I would be interested in the numbers pre/post-patch. 

We can obtain this with ”casperf -b 512 -q 64 ’malloc:///disk0?size_mb=100’” I think that even with debug build it would already show an improvement

**Comment by daniel-shuy** [comment]
@tiagolobocastro @gila thanks, I now understand that `poll_times()` and `schedule()` are called from the same thread. I've dropped the `AtomicBool` change.

```
In other words, your original code was what we want I think, just need to check that the code generated is correct 👍
```
Sorry I'm kinda new to Rust, how can I see the generated code?

```
The original patch -- i.e. without the atomic bool was indeed correct. I would be interested in the numbers pre/post-patch.

We can obtain this with ”casperf -b 512 -q 64 ’malloc:///disk0?size_mb=100’” I think that even with debug build it would already show an improvement
```
Unfortunately I am unable to fully compile and run the project, it fails with "version `GLIBC_2.32' not found" (with `nix-shell`).

**Comment by tiagolobocastro** [comment]
IIRC @blaisedias hit similar glibc issue recently?

**Comment by gila** [comment]
We recently bumped nix packages so perhaps there is some overlap. Could you exit (all of them!) nix-shell and run the garbage collector and re-enter the shell?

**Comment by daniel-shuy** [comment]
> We recently bumped nix packages so perhaps there is some overlap. Could you exit (all of them!) nix-shell and run the garbage collector and re-enter the shell?

Thanks, that worked.

However, when I run `./casperf -b 512 -q 64 'malloc:///disk0?size_mb=100'`, it fails with:
`EAL: No available hugepages reported in hugepages-2048kB`

After running `echo 512 | sudo tee  /sys/kernel/mm/hugepages/hugepages-2048kB/nr_hugepages`, `casperf` fails with:
`EAL: 512 hugepages of size 2097152 reserved, but no mounted hugetlbfs found for that size`

I then tried running:
```
mkdir -p /mnt/huge
sudo mount -t hugetlbfs nodev /mnt/huge
```
and it now fails with:
`EAL: Couldn't get fd on hugepage file`

I'm not sure if this has to do with me running on WSL.

**Comment by daniel-shuy** [comment]
I tried the Docker approach as well, but running `cargo`/`rustc` throws `SegmentationFault`

**Comment by jkryl** [comment]
I'm closing this PR as it hasn't been active for quite some time already. If the author wishes to continue in this effort, feel free to re-open. thanks

**Pull request closed without merging**